### PR TITLE
box: disallow concurrent invocation of `box_raft_try_promote`

### LINKED ISSUE #11703: replication: initial promote attempt after bootstrap leads to assertion failure
Tarantool version: 3.5.0-entrypoint-132-g8c7cfb1cd0

## Reproduce files

`i.lua`:

```lua
#!/usr/bin/env tarantool

local fio = require('fio')
local fiber = require('fiber')

local port = tonumber(arg[1])
assert(port ~= nil, port)
fio.mkdir(port)
if port == 3301 then
    fiber.new(function()
        fiber.sleep(2)
        box.ctl.make_bootstrap_leader({graceful = true})
        while box.info.status ~= 'running' do
            fiber.sleep(0.001)
        end
        box.ctl.promote()
        box.cfg({read_only = false})
        box.ctl.wait_rw()
        box.schema.user.grant('guest', 'super')
    end)
end
box.cfg({
    read_only = true,
    work_dir = tostring(port),
    listen = port,
    replication = {3301, 3302, 3303},
    election_mode = 'manual',
    bootstrap_strategy = 'supervised',
})
```

`Procfile`:

```procfile
i-3301: ./src/tarantool i.lua 3301
i-3302: ./src/tarantool i.lua 3302
i-3303: ./src/tarantool i.lua 3303
```

## Reproduce steps

```sh
$ rm -r 3301 3302 3303
$ honcho start
```

## Result

```
tarantool: ./src/box/box.cc:6286: void box_cfg_xc(): Assertion `rc == 0' failed.
```

The relevant code (the assertion is on the 6286 line):

https://github.com/tarantool/tarantool/blob/8c7cfb1cd0b3f5ce1d5b181403f68f7aaf657d9a/src/box/box.cc#L6265-L6288

### LINKED ISSUE #11708: replication: box.ctl.promote() fails with assertion
Tarantool version: 3.5.0-entrypoint-132-g8c7cfb1cd0

## Reproduce files

`i.lua` (same as in #11703 except `fiber.new(box.ctl.promote)`):

```lua
#!/usr/bin/env tarantool

local fio = require('fio')
local fiber = require('fiber')

local port = tonumber(arg[1])
assert(port ~= nil, port)
fio.mkdir(port)
if port == 3301 then
    fiber.new(function()
        fiber.sleep(2)
        box.ctl.make_bootstrap_leader({graceful = true})
        while box.info.status ~= 'running' do
            fiber.sleep(0.001)
        end
        fiber.new(box.ctl.promote)
        box.ctl.promote()
        box.cfg({read_only = false})
        box.ctl.wait_rw()
        box.schema.user.grant('guest', 'super')
    end)
end
box.cfg({
    read_only = true,
    work_dir = tostring(port),
    listen = port,
    replication = {3301, 3302, 3303},
    election_mode = 'manual',
    bootstrap_strategy = 'supervised',
})
```

`Procfile` (same as in #11703):

```procfile
i-3301: ./src/tarantool i.lua 3301
i-3302: ./src/tarantool i.lua 3302
i-3303: ./src/tarantool i.lua 3303
```

## Reproduce steps

```sh
$ rm -r 3301 3302 3303
$ honcho start
<..wait for gh-11703 crash..>
$ honcho start
```

## Result

```
tarantool: ./src/lua/error.c:123: luaT_error_at: Assertion `e != NULL' failed.
```

The stacktrace:

```c
(gdb) bt
#0  0x00007ffff5015e6c in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff4fc7612 in raise () from /lib64/libc.so.6
#2  0x00007ffff4fb04ed in abort () from /lib64/libc.so.6
#3  0x00007ffff4fb0415 in __assert_fail_base.cold () from /lib64/libc.so.6
#4  0x00007ffff4fc0192 in __assert_fail () from /lib64/libc.so.6
#5  0x0000555555853147 in luaT_error_at (L=0x402fca88, level=1) at /home/alex/p/tarantool-meta/tarantool/src/lua/error.c:123
#6  0x00005555558531c6 in luaT_error (L=0x402fca88) at /home/alex/p/tarantool-meta/tarantool/src/lua/error.c:135
#7  0x0000555555812083 in lbox_ctl_promote (L=0x402fca88) at /home/alex/p/tarantool-meta/tarantool/src/box/lua/ctl.c:106
#8  0x00005555558e1cf7 in lj_BC_FUNCC () at buildvm_x86.dasc:811
#9  0x00005555558ef6e3 in lua_pcall (L=0x402fca88, nargs=0, nresults=-1, errfunc=0) at /home/alex/p/tarantool-meta/tarantool/third_party/luajit/src/lj_api.c:1173
#10 0x000055555584d1f3 in luaT_call (L=0x402fca88, nargs=0, nreturns=-1) at /home/alex/p/tarantool-meta/tarantool/src/lua/utils.c:708
#11 0x00005555558421c7 in lua_fiber_run_f (ap=0x7ffff4410a30) at /home/alex/p/tarantool-meta/tarantool/src/lua/fiber.c:452
#12 0x0000555555593434 in fiber_cxx_invoke(fiber_func, typedef __va_list_tag __va_list_tag *) (f=0x55555584210e <lua_fiber_run_f>, ap=0x7ffff4410a30) at /home/alex/p/tarantool-meta/tarantool/src/lib/core/fiber.h:1324
#13 0x000055555587e017 in fiber_loop (data=0x0) at /home/alex/p/tarantool-meta/tarantool/src/lib/core/fiber.c:1167
#14 0x0000555555bd3568 in coro_init () at /home/alex/p/tarantool-meta/tarantool/third_party/coro/coro.c:108
```

The relevant code:

https://github.com/tarantool/tarantool/blob/8de5bae07a1a2b35d04be7d3dcace6fc826fafb1/src/box/lua/ctl.c#L102-L108

https://github.com/tarantool/tarantool/blob/8de5bae07a1a2b35d04be7d3dcace6fc826fafb1/src/lua/error.c#L132-L136

https://github.com/tarantool/tarantool/blob/8de5bae07a1a2b35d04be7d3dcace6fc826fafb1/src/lua/error.c#L119-L130

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
Closes #11703
Closes #11708

NO_DOC=<bugfix>

### CODE REVIEW

File: test/replication-luatest/gh_11708_manual_leader_election_concurrent_promote_crashes_test.lua
Comment (Gerold103) [inline test/replication-luatest/gh_11708_manual_leader_election_concurrent_promote_crashes_test.lua:1]: Tests need to have `gh_xxxx` prefix telling the ticket number. In this case I guess this would be 2 numbers as `gh_xxxx_yyyy_` or any one of them.

File: src/box/box.cc [hunk 1]
```diff
@@ -6286,16 +6285,19 @@ box_cfg_xc(void)
...
 		/*
 		 * When the cluster is just bootstrapped and this instance is a
 		 * leader, it makes no sense to wait for a leader appearance.
 		 * There is no one. Moreover this node *is* a leader, so it
 		 * should take the control over the situation and start a new
 		 * term immediately.
 		 */
+		is_in_box_promote = true;
 		int rc = box_raft_try_promote();
+		is_in_box_promote = false;
```
Comment (Gerold103) [inline src/box/box.cc]: From the commit message I understood that the problem was in `box_raft_try_promote()`'s trigger stealing diag from the raft's worker fiber, and this led to issues when there are multiple triggers? But in all cases of `diag_set()` in `box_raft_try_promote_f()` I see that it is later moved. So the `diag_move()` always moves the error which was just installed in the same function. How come the triggers are conflicting?

Is it because of the first `if (raft->state == RAFT_STATE_LEADER) goto done` which sets no error, but does `diag_move()`?

It seems to me that this is the main bug here. Some shady business happening with diags. It got here workarounded with an external global flag, but the underlying issue is still there, apparently.

Wouldn't it all just work if we would be careful with diag in `box_raft_try_promote_f()`?

The main problem I see now is that `box_raft_try_promote()` is not self-sufficient anymore. It now has to be always called with this global flag checked and set, which indicates that this function can't be "public" (can't be in any header).

File: changelogs/unreleased/manual-leader-election-concurrent-promote-crashes.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## bugfix/replication
+
+* Fixed two related bugs when concurrent `box.ctl.promote` invocations with
+  `box.cfg.election_mode = 'manual'` would crash during (gh-11703) or after
+  (gh-11708) bootstrap.
```
Comment (Gerold103) [inline changelogs/unreleased/manual-leader-election-concurrent-promote-crashes.md]: I somehow can't understand what is happening. Both tickets are only different in how many `box.ctl.promote()`s they do. They are not different in when they call the promotes. Both seem to be doing it after the bootstrap. They are both firstly waiting while `box.info.status ~= 'running'` and only then do the promotes.

Home come 11703 is said to happen during bootstrap? Wouldn't then `info.status` be not `running`, preventing `box.ctl.promote()` from being called?

Comment (CuriousGeorgiy) [reply changelogs/unreleased/manual-leader-election-concurrent-promote-crashes.md]: Sorry, I mixed up the concepts of bootstrap and configuration. I was referring to this piece of code (which happens after bootstrap):
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/box.cc#L6288-L6314

Reworded: bootstrap -> server configuration via `box.cfg`

File: test/replication-luatest/gh_11708_manual_leader_election_concurrent_promote_crashes_test.lua
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_11708_manual_leader_election_concurrent_promote_crashes_test.lua:1]: Added the `gh_11708` prefix as I think #11703 is just a corner case of #11708.

File: src/box/box.cc (hunk 1 above)
Comment (CuriousGeorgiy) [reply src/box/box.cc]: > From the commit message I understood that the problem was in box_raft_try_promote()'s trigger stealing diag from the raft's worker fiber.

No, it is stealing the diag from the fiber that is concurrently running `box.ctl.promote`, since the `raft_on_update` trigger is executed synchronously with the following call chains:
1. `raft_schedule_broadcast` <- `raft_stop_candidate` <- `raft_set_candidate` <- `raft_restore` <- `raft_sm_schedule_new_term` <- `raft_promote` <- `box_raft_try_promote`
2. `raft_schedule_broadcast` <- `raft_stop_candidate` <- `raft_set_candidate` <- `raft_restore` <- `box_raft_try_promote`

The following sequence of actions happens:
1. First fiber sets its `box_raft_try_promote_f` trigger and starts waiting to become leader:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L538-L545
2. Second fiber concurrently invokes `box_raft_try_promote`, disables the `is_candidate` state and invokes `box_rafter_promote_f`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/lib/raft/raft.c#L911-L915
3. In `box_rafter_promote_f` we realize that we are no longer a candidate and set a `diag`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L497-L505
4. In  `box_rafter_promote_f` we move the `diag` to the trigger `ctx`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L516
5. Second fiber sets its `box_raft_try_promote_f` trigger and starts waiting to become leader.
6. First fiber wakes up, realizes its trigger finished executing but it did not become a leader and moves the `diag` from the trigger context to its `fiber` state:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L562-L563
7. First fiber executes `raft_restore`, disables the `is_candidate` state and invokes `box_rafter_promote_f`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L568
8. In `box_rafter_promote_f` we realize that we are no longer a candidate and overwrite the `diag` **in the first `fiber`'s state**, which already had a `diag` set from step 6.
9. In `box_rafter_promote_f` we move the `diag` from the `fiber` state to the trigger `ctx`, resetting the second `fiber` state's `diag`.
10. The first fiber returns an error from `box_raft_try_promote` and either tries to access the `diag` which was stolen by the first fiber and is set to `NULL`, which causes a crash:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/lua/ctl.c#L105-L106
Or panics in `box.cfg`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/box.cc#L6303-L6312

> Is it because of the first `if (raft->state == RAFT_STATE_LEADER) goto done` which sets no error, but does `diag_move()`?

No, this is completely unrelated, since the concurrently invoked promotes can never succeed do to the conflicting `raft_restore` calls.

> It seems to me that this is the main bug here. Some shady business happening with diags. It got here workarounded with an external global flag, but the underlying issue is still there, apparently.

I don't think so. The main bug is related to concurrently invoking `box_raft_try_promote` and modifying `raft` state in a conflicting way because of `raft_restore`.

> Wouldn't it all just work if we would be careful with diag in `box_raft_try_promote_f()`?

This would still cause `box.cfg` to panic, and I don't think that the current behavior that causes all concurrently executing promotes to fail is correct.

> The main problem I see now is that box_raft_try_promote() is not self-sufficient anymore. It now has to be always called with this global flag checked and set, which indicates that this function can't be "public" (can't be in any header).

I just implemented an alternative solution to encapsulate the `is_in_box_promote` manipulation inside `box_raft_try_promote` like we do for `box_promote_qsync`:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/box.cc#L3306-L3312

I don't like the crudge with having to set `is_in_box_promote` to `false` in `box_promote` before calling `box_raft_try_promote`. Instead I have added a static `is_in_raft_promote` flag to `box_raft_try_promote` to encapsulate the guard.

It also seems that we can rely on the fact that we do not yield in between `title` and `box_raft_try_promote` in `box_cfg_xc`, so the promote invoked in `box_cfg_xc` is always going to be the first promote among concurrent invocations, so we don't need to check `is_in_box_promote`.

Comment (Gerold103) [reply src/box/box.cc]: I don't get the step 8:

> In the trigger set by the first fiber we realize that we are no longer a candidate and set a diag in the second fiber's state, which already has a diag set from step 6.

The fiber's context `struct box_raft_watch_ctx ctx` is given to the fiber's trigger as a part of that trigger: `trigger_create(&trig, box_raft_try_promote_f, &ctx, NULL);`. In `box_raft_try_promote_f()` we take this `ctx`. The contexts of different fibers have no way to get messed into different triggers. How does that happen?

File: src/box/box.cc [hunk 3]
```diff
@@ -6291,7 +6290,7 @@ box_cfg_xc(void)
 	if (!is_bootstrap_leader) {
 		replicaset_sync();
 	} else if (box_election_mode == ELECTION_MODE_CANDIDATE ||
-		   box_election_mode == ELECTION_MODE_MANUAL) {
+		    box_election_mode == ELECTION_MODE_MANUAL) {
 		/*
```
Comment (Gerold103) [inline src/box/box.cc:6295]: Unnecessary change.

File: src/box/box.cc (hunk 1 above)
Comment (CuriousGeorgiy) [reply src/box/box.cc]: > The contexts of different fibers have no way to get messed into different triggers. How does that happen?

We interfere with the state of the fiber that is executing the trigger by calling `diag_set` and `diag_move` inside the trigger:
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L497-L505
https://github.com/tarantool/tarantool/blob/1c0f54da3b8c98dd87405f42ba7a15fd12888d2f/src/box/raft.c#L516

Potentially the problem could be fixed by avoiding this interference and setting the diag directly on `ctx.diag`. But even if we fix the crash this way, the `panic` problem during `box.cfg` will still be there.

Comment (Gerold103) [reply src/box/box.cc]: From what I have investigated, the problem isn't the step 8, because it couldn't happen as you describe it. It is just not possible. For the 2 contexts to get mixed by 2 triggers they would have to be connected at least somehow, but they aren't.

Here is what I found. The trigger `box_raft_try_promote_f()` is executed 2 times for the same trigger. I removed your `is_in_raft_promote` and added `assert(!ctx->is_done)` to the beginning of `box_raft_try_promote_f()`. And it failed. So apparently, the trigger due to concurrent restore/broadcast (what you describe above in steps 1-7) is just done twice. First time it sets "done" and wakes the original fiber up. But when invoked second time, it might change its result. Later the original fibers wakes up and sees that it is not a leader atm, tries to get the diag, but the diag is empty. And we don't check that.

I am wondering if this scenario of the trigger being executed twice and changing its own state can still break something even after your fix. For example, an instance calls the promote, then receives a couple of updates from a remote instance, causing the trigger being done twice again. Can that happen?

Here was my patch:
```Diff
diff --git a/src/box/raft.c b/src/box/raft.c
index 201ee90904..9b2af62fdb 100644
--- a/src/box/raft.c
+++ b/src/box/raft.c
@@ -464,6 +464,7 @@ box_raft_try_promote_f(struct trigger *trig, void *event)
 	struct raft *raft = event;
 	assert(raft == box_raft());
 	struct box_raft_watch_ctx *ctx = trig->data;
+	assert(!ctx->is_done);
 
 	if (raft->state == RAFT_STATE_LEADER)
 		goto done;
@@ -523,8 +524,7 @@ box_raft_try_promote(void)
 {
 	static bool is_in_raft_promote = false;
 	if (is_in_raft_promote) {
-		diag_set(ClientError, ER_IN_ANOTHER_PROMOTE);
-		return -1;
+		say_info("nop");
 	}
 	struct raft *raft = box_raft();
 	is_in_raft_promote = true;
```
And your test failed on this new assert. So the context wasn't stolen AFAIU. It was overwritten by the same trigger. (Doing just return when already done didn't help - the state is still stale by the time the original fiber wakes up.)

Comment (CuriousGeorgiy) [reply src/box/box.cc]: > From what I have investigated, the problem isn't the step 8, because it couldn't happen as you describe it. It is just not possible. For the 2 contexts to get mixed by 2 triggers they would have to be connected at least somehow, but they aren't.

I don't understand what "context mix up" you are talking about, I never mentioned this notion.

> Later the original fibers wakes up and sees that it is not a leader atm, tries to get the diag, but the diag is empty. And we don't check that.

No, this is not the case. The original fiber correctly sets its `diag` in step 6. Then, when executing `raft_restore` and invoking the `box_raft_try_promote_f` trigger again, the fiber's `diag` effectively gets set to `NULL` because of the `diag_set` and `diag_move` happening in the `box_raft_try_promote_f` trigger. And then the fiber crashes when we try to throw its `diag` which was overwritten to `NULL`.

> And your test failed on this new assert. So the context wasn't stolen AFAIU. It was overwritten by the same trigger. (Doing just return when already done didn't help - the state is still stale by the time the original fiber wakes up.)

Of course it failed — you disabled the `is_in_raft_promote` with your patch. If I leave the assertion you have added but omit your `is_in_raft_promote` guard change, the test passes.

> I am wondering if this scenario of the trigger being executed twice and changing its own state can still break something even after your fix. For example, an instance calls the promote, then receives a couple of updates from a remote instance, causing the trigger being done twice again. Can that happen?

Again, the problem is not with the trigger changing its own state, the problem is with the trigger changing the state of the fiber that is executing it with `diag_set` and `diag_move`. We can safeguard the code by avoiding using `diag_set`, but I don't see any other pitfalls.

File: src/box/box.cc (hunk 3 above)
Comment (CuriousGeorgiy) [reply src/box/box.cc:6295]: Sorry, fixed. Weird that it slipped through `checkpatch`.

File: src/box/box.cc (hunk 1 above)
Comment (Gerold103) [reply src/box/box.cc]: > the problem is with the trigger changing the state of the fiber that is executing it with diag_set and diag_move

I don't get this. The trigger doesn't touch the fiber besides doing `fiber_wakeup()`. The diag is local to `box_raft_watch_ctx`. It is a member of this struct. The trigger doesn't change any fiber's diag directly.

> Of course it failed — you disabled the is_in_raft_promote with your patch

This isn't the point. The point is that the trigger gets executed more than once. Your patch is masking it, but the trigger still can get executed twice.

I am afraid we are stuck here. If you want, you can try to show me this on a call with screen share. Otherwise I don't understand how the triggers are stealing/overwriting each other's diags or fibers' diags.

Comment (Gerold103) [reply src/box/box.cc]: @Serpentian , can you please help with another pair of eyes? I think am missing something obvious 😭.

Comment (CuriousGeorgiy) [reply src/box/box.cc]: Or @sergepetrenko

Comment (Serpentian) [reply src/box/box.cc]: The first thing I've tried, was adding assertions to error in diag with the following patch:

```diff
diff --git a/src/box/raft.c b/src/box/raft.c
index e0b0b65a05..8ef53da814 100644
--- a/src/box/raft.c
+++ b/src/box/raft.c
@@ -565,7 +565,9 @@ box_raft_try_promote(void)
 		diag_set(TimedOut);
 	}
 
+	assert(diag_last_error(diag_get()) != NULL);
 	raft_restore(raft);
+	assert(diag_last_error(diag_get()) != NULL);
 	return -1;
 }
```

I ran the `test_concurrent_promote_after_box_cfg` test. Only assertion on the last line failed, the first one passed, this сonfirms the hypothesis of Georgiy, that after `raft_restore` the diag is somehow overwritten.

>  Later the original fibers wakes up and sees that it is not a leader atm, tries to get the diag, but the diag is empty. And we don't check that.

So, this cannot happen, error is there, when we go out of the loop and check for leader in the `box_raft_try_promote`, but it's not there after `raft_restore`, which executes the trigger of the promote in another fiber.

----

After that I just took GDB, ran tarantool there with the test of Georgiy and have seen the following:

```
tarantool> box.cfg{election_mode = 'manual'}
tarantool> local fiber = require('fiber'); local f = fiber.new(box.ctl.promote); f:set_joinable(true); local ok = pcall(box.ctl.promote); t.assert(ok); local err; ok, err = f:join()
```

Here's the log of the gdb, @Gerold103. I think, u suppose, that we somehow overwrite the diag in the `ctx`, but this is not what happens. We overwrite `fiber()->diag->last` because it may happen, that we set the error after exiting from the `raft_try_promote_f` and then overwrite it in the `raft_restore`, when executing `raft_promote_f`, which is ran in the same fiber.

[gdb.txt](https://github.com/user-attachments/files/23103248/gdb.txt)

Here's the example of patch, which also fixes the crash of the `error`, but of course, it doesn't fix the panic in `box.cfg`, since it relies on the fact, that `raft_try_promote` never fails:

```diff
diff --git a/src/box/raft.c b/src/box/raft.c
index e0b0b65a05..7bd3370060 100644
--- a/src/box/raft.c
+++ b/src/box/raft.c
@@ -557,6 +557,7 @@ box_raft_try_promote(void)
 		return 0;
 	}
 
+	raft_restore(raft);
 	if (fiber_is_cancelled()) {
 		diag_set(FiberIsCancelled);
 	} else if (ctx.is_done) {
@@ -565,7 +566,6 @@ box_raft_try_promote(void)
 		diag_set(TimedOut);
 	}
 
-	raft_restore(raft);
 	return -1;
 }
 
``` 

At the end, Georgiy is right here, we overwrite the diag of the current fiber

Comment (Serpentian) [reply src/box/box.cc]: We must be sure, that none of the `raft_on_update` triggers may cause the overwrite of the fiber's diag. We should either move `raft_restore` higher, as I proposed, or at least add assertions to the `raft_schedule_broadcast`, that the fiber's diag cannot be changed, when executing triggers (but IMHO, that's normal behavior and the first solution is better, though I don't know, will it properly work)

But for fixing the panic in box.cfg we should either prohibit simultaneous promotes, as Georgiy proposes, or retry the `box_raft_try_promote` call there?

Comment (CuriousGeorgiy) [reply src/box/box.cc]: > I don't get this. The trigger doesn't touch the fiber besides doing `fiber_wakeup()`. The diag is local to `box_raft_watch_ctx`. It is a member of this struct. The trigger doesn't change any fiber's diag directly.

@Gerold103 please take a look at `diag_set` and `diag_get` definitions:
https://github.com/tarantool/tarantool/blob/0628d91cd76be45cfe0623d955237ca4c7125e0c/src/lib/core/diag.c#L152-L156
https://github.com/tarantool/tarantool/blob/0628d91cd76be45cfe0623d955237ca4c7125e0c/src/lib/core/diag.h#L470-L470

Comment (Gerold103) [reply src/box/box.cc]: @CuriousGeorgiy , the definition of `diag_set()` and I knew, this wasn't confusing.

Thanks, @Serpentian , I see now. With the specific code sample in and the trick with `raft_restore()` it became clear.

I still think this is dangerous then to just leave 1) the trigger able to get executed more than once, 2) `raft_restore()`, as noted by Nikita, to be left after all `diag_set()` calls. But I can't prove right anyway with any tests how it would break, so the current solution should be ok. Thanks, Georgy, for the patch!

**Comment by Serpentian** [comment]
> We must be sure, that none of the raft_on_update triggers may cause the overwrite of the fiber's diag. We should either move raft_restore higher, as I proposed, or at least add assertions to the raft_schedule_broadcast, that the fiber's diag cannot be changed, when executing triggers (but IMHO, that's normal behavior and the first solution is better, though I don't know, will it properly work)

I'd like to note, that we should somehow cover that place, so that we don't introduce new triggers tomorrow, that'll clear the diag. Assertions in the `raft_schedule_broadcast` should be enough.

File: src/box/raft.c [hunk 4]
```diff
@@ -566,6 +573,7 @@ box_raft_try_promote(void)
 	}
 
 	raft_restore(raft);
```
Comment (Serpentian) [inline src/box/raft.c]: Can you consider moving `raft_restore` higher? The alternative solution would be adding assertions to the `raft_schedule_broadcast`. I wanna be sure, that tomorrow we won't introduce new `raft_on_update` triggers, which don't depend on `box_raft_try_promote`.  This will break it all again and reincarnate that bug

File: src/box/raft.c [hunk 5]
```diff
@@ -521,7 +521,13 @@ box_raft_try_promote_f(struct trigger *trig, void *event)
 int
 box_raft_try_promote(void)
 {
+	static bool is_in_raft_promote = false;
+	if (is_in_raft_promote) {
+		diag_set(ClientError, ER_IN_ANOTHER_PROMOTE);
```
Comment (Serpentian) [inline src/box/raft.c:529]: How can we be sure, that this error won't be returned to the box.cfg and then `assert(rc == 0)` will fail again. For that we need `box.ctl.promote` to be started before the box.cfg's promote. Is it possible.

It seems to me, that the only possible solution there is retry of `box_raft_try_promote`

P.S. If you apply retry and move `raft_restore` higher, then we don't need to prohibit simultaneous `box_raft_try_promote`

Comment (CuriousGeorgiy) [reply src/box/raft.c:529]: The situation you describe is not possible, `box_raft_try_promote` will always be called from `box.cfg` first. This is because between setting `is_box_configured` to `true` (which is a pre-condition for `box_promote` execution) and calling `box_raft_try_promote` we do not yield:
https://github.com/tarantool/tarantool/blob/cc77a6ead00b9a07233f302dabae994fc79dde21/src/box/box.cc#L6254-L6303

I dislike adding retry logic to `box.cfg`, since it will complicate the code and defining the retrying semantics is not clear (how many times should we retry before panicking? with what interval should we retry?). I would prefer to maintain the invariant we have right now that `box.cfg` is always the first to call `box_raft_try_promote`.

To be on the safer side, I have moved the `is_box_configured` update right before the `box_raft_try_promote`.

File: src/box/box.cc [hunk 6]
```diff
@@ -6286,6 +6284,7 @@ box_cfg_xc(void)
 	if (dd_version_id > version_id(2, 10, 1))
 		txn_limbo_filter_enable(&txn_limbo);
 
+	is_box_configured = true;
```
Comment (sergepetrenko) [inline src/box/box.cc:6286]: Shouldn't the `box_broadcast_ballot` which was formerly below the `is_box_configued` setting be moved here as well?
I don't see any immediate consequences, but it doesn't look safe that we don't broadcast `ballot->is_booted` anymore.

Comment (CuriousGeorgiy) [reply src/box/box.cc:6286]: Makes sense, moved `box_broadcast_ballot` too.

Comment (sergepetrenko) [reply src/box/box.cc:6286]: Thanks!

**Review by sergepetrenko — APPROVED** [review]
LGTM!
